zerver/lib/markdown/__init__.py

```diff
@@ -1595,10 +1595,55 @@ def __init__(self, compiled_re: Pattern, md: markdown.Markdown) -> None:
 
 
 class AutoLink(CompiledPattern):
+    def shorten_links(self, href: str) -> Optional[str]:
+        parts = urllib.parse.urlparse(href)
+        scheme, netloc, path, params, query, fragment = parts
+        text = None
+        if netloc in ["github.com", "www.github.com"]:
+            text = self.shorten_github_links(path, fragment)
+        return text
+
+    def shorten_github_links(self, path: str, fragment: str) -> Optional[str]:
+        # Split the path to extract our 4 variables.
+        #
+        # To do it cleanly without multiple if branches based on which of these
+        # variables are present, we here add a list of [None, None, None...]
+        # to the result of path.split, which at worst can be []. We also remove
+        # the first empty string we'd get from '/foo/bar'.split('/').
+        #
+        # Example path: '/foo/bar' output: ['foo', 'bar', None, None]
+        #         path: ''         output: [None, None, None, None]
+        organisation, repository, artifact, value, remaining_path = (
+            path.split("/", 5)[1:] + [""] * 5
+        )[:5]
```

Comment:
You’ve commented `None` but coded `""`.

Many GitHub URLs don’t obey this scheme, e.g. https://github.com/settings/profile, https://github.com/sponsors/accounts, https://github.com/marketplace/circleci, https://github.com/account/choose, https://github.com/about/press.

I agree with Tim that it’s dangerous to try to parse parts of URLs that we don’t fully understand. We should have a very restrictive parser that works for a fixed collection of common patterns and errs on the side of not shortening.

Reply:
@andersk, Ah! Sorry, forget to correct it.

About the other part, we now agree only to shorten links related to issues and PRs only.
More info in [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Automatically.20shorten.20Github.20URLs.20.2311895).